### PR TITLE
[PS-11868] Require key for enc string decryption

### PR DESCRIPTION
**Pull request by MGibson1**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-11868

## 📔 Objective

Adds a new interface to enc string that requires a key and `EncryptService` to encrypt, rather than trying to determine keys and grabbing services from a global container.

Additionally uses this interface through the base `Domain` class that _also_ requires a key and encrypt service. Additionally, this new interface is type safe to decrypt only `EncString`s on the object and avoids mutation of a passed in object.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: libs/common/src/vault/models/domain/folder.ts
Comment (MGibson1) [inline libs/common/src/vault/models/domain/folder.ts:1]: These folder updates are example usages. We can remove from this PR to limit scope.

File: libs/common/src/platform/models/domain/domain-base.ts [hunk 1]
```diff
@@ -80,4 +90,60 @@ export default class Domain {
...
+   */
+  protected async decryptObjWithKey<
+    TThis extends Domain,
+    const TEncryptedKeys extends EncStringKeys<TThis>,
+  >(
+    this: TThis,
+    encryptedProperties: TEncryptedKeys[],
+    key: SymmetricCryptoKey,
+    encryptService: EncryptService,
+    _: Constructor<TThis> = this.constructor as Constructor<TThis>,
```
Comment (MGibson1) [inline libs/common/src/platform/models/domain/domain-base.ts:114]: The type inferrer has a lot of trouble with `this` type on derived classes. It seems to always resolve to the parent class when using in the derived class. This pattern allows us to get by that by providing a constructor as an optional 5th argument. [See the `Folder` changes for an example of its value.](https://github.com/bitwarden/clients/pull/10981/files#diff-26dbab7659d26dca711d2667af220b314da6d2c54ae05807d9cc285fabe22324R54)

File: libs/common/src/platform/models/domain/enc-string.ts [hunk 2]
```diff
@@ -167,11 +170,24 @@ export class EncString implements Encrypted {
...
+      this.decryptedValue = DECRYPT_ERROR;
     }
     return this.decryptedValue;
   }
 
+  async decryptWithKey(key: SymmetricCryptoKey, encryptService: EncryptService) {
+    try {
+      if (key == null) {
+        throw new Error("No key to decrypt EncString");
+      }
```
Comment (justindbaur) [inline libs/common/src/platform/models/domain/enc-string.ts:182]: Should this be moved outside of the `try`? Not providing the key should be a developer problem right?

**Review by justindbaur — COMMENTED** [review]
Just one question on if we should throw an error sooner in one place.

File: libs/common/src/platform/models/domain/enc-string.ts (hunk 2 above)
Comment (justindbaur) [reply libs/common/src/platform/models/domain/enc-string.ts:182]: Okay fair enough, wasn't sure if we wanted to go with a runtime error until we can get strict null but I agree that's the desire.

**Review by justindbaur — APPROVED** [review]

**Review by gbubemismith — APPROVED** [review]
This looks good 🚀

**Pull request merged**